Dockerfile

```diff
@@ -13,46 +13,352 @@
+FROM airflow-apt-deps as airflow-ci-apt-deps
+
+SHELL ["/bin/bash", "-xc"]
+
+ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64/
+
+# Note missing directories on debian-stretch https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=863199
+RUN mkdir -pv /usr/share/man/man1 \
+    && mkdir -pv /usr/share/man/man7 \
+    && apt-get update \
+    && apt-get install --no-install-recommends -y \
+      lsb-release \
+      gnupg \
+      dirmngr \
+      openjdk-8-jdk \
+      vim \
+      wget \
+      tmux \
+      less \
+      unzip \
+      ldap-utils \
+      postgresql-client \
+      sqlite3 \
+      krb5-user \
+      openssh-client \
+      openssh-server \
+      python-selinux \
+      sasl2-bin \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+RUN export DEBIAN_FRONTEND=noninteractive \
+# gpg: key 5072E1F5: public key "MySQL Release Engineering <mysql-build@oss.oracle.com>" imported
+    && key='A4A9406876FCBD3C456770C88C718D3B5072E1F5' \
+    && export GNUPGHOME="$(mktemp -d)" \
+    && for keyserver in $(shuf -e \
+			ha.pool.sks-keyservers.net \
+			hkp://p80.pool.sks-keyservers.net:80 \
+			keyserver.ubuntu.com \
+			hkp://keyserver.ubuntu.com:80 \
+			pgp.mit.edu) ; do \
+		  gpg --keyserver $keyserver --recv-keys "$key" && break || true ; \
+	   done \
+    && gpg --export "$key" > /etc/apt/trusted.gpg.d/mysql.gpg \
+	&& gpgconf --kill all \
+	rm -rf "$GNUPGHOME"; \
+	apt-key list > /dev/null \
+    && echo "deb http://repo.mysql.com/apt/ubuntu/ trusty mysql-5.7" | tee -a /etc/apt/sources.list.d/mysql.list \
+    && apt-get update \
+    && MYSQL_PASS="secret" \
+    && debconf-set-selections <<< "mysql-community-server mysql-community-server/data-dir select ''" \
+    && debconf-set-selections <<< "mysql-community-server mysql-community-server/root-pass password $MYSQL_PASS" \
+    && debconf-set-selections <<< "mysql-community-server mysql-community-server/re-root-pass password $MYSQL_PASS" \
+    && apt-get install --no-install-recommends -y mysql-client libmysqlclient-dev \
+    && apt-get autoremove -yqq --purge \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+ENV HADOOP_DISTRO=cdh
+ENV HADOOP_VERSION=2.6.0 HADOOP_HOME=/tmp/hadoop-${HADOOP_DISTRO} HIVE_HOME=/tmp/hive
+
+RUN  mkdir -pv ${HADOOP_HOME} && \
+     mkdir -pv ${HIVE_HOME}  && \
+     mkdir /tmp/minicluster  && \
+     mkdir -pv /user/hive/warehouse && \
+     chmod -R 777 ${HIVE_HOME} && \
+     chmod -R 777 /user/
+
+# Install Hadoop
+# --absolute-names is a work around to avoid this issue https://github.com/docker/hub-feedback/issues/727
+RUN cd /tmp && \
+    wget -q https://archive.cloudera.com/cdh5/cdh/5/hadoop-${HADOOP_VERSION}-cdh5.11.0.tar.gz && \
+    tar xzf hadoop-${HADOOP_VERSION}-cdh5.11.0.tar.gz --absolute-names --strip-components 1 -C ${HADOOP_HOME} && \
+    rm hadoop-${HADOOP_VERSION}-cdh5.11.0.tar.gz
+
+# Install Hive
+RUN cd /tmp && \
```

Comment:
I know that you just moved this from the original code, but now we're at it, maybe we can change the wget to curl, and then we can also drop wget from the installation: `curl -s https://archive.cloudera.com/cdh5/cdh/5/hadoop-${HADOOP_VERSION}-cdh5.11.0.tar.gz | tar xzf - --absolute-names --strip-components 1 -C ${HADOOP_HOME}`

Reply:
Sure. Will keep it open :)

Reply:
@fokko - I added the linter you suggested, and fixed all the inconsistences found and problems reported by the linter. It's pretty good one (the linter). I've added linting of Dockerfile as pre-test stage in Travis as well as I am doing it automatically before build with Airflow Breeze https://github.com/apache/airflow/pull/4932  :).

I had to however disable some of the warnings - specifically:

  - DL3006: https://github.com/hadolint/hadolint/wiki/DL3006 "Always tag the version of an image explicitly." - since we have dynamic base images in our Dockerfile the linter is not smart enough to figure out what we are doing.... I have to disable it 

  - DL3008: https://github.com/hadolint/hadolint/wiki/DL3008 - "Pin versions in apt get install."  I agree it's a bad practice but It seems Airflow leaves in a bit more dynamic world and we prefer to always install the "latest possible versions" of both APT and PIP. This we might discuss when we get back to "how to handle dependencies" community discussion - but it's a separate AIP I believe. 

  - DL3005: https://github.com/hadolint/hadolint/wiki/DL3005 "Do not use apt-get upgrade or dist-upgrade." - same as above

  - DL3013: https://github.com/hadolint/hadolint/wiki/DL3013  "Pin versions in pip". - same story

I like the linter. It pointed me to a number of problems I did not realise (for example using sudo - now I replaced it with "gosu airflow" when needed and put USER airflow at the end of the Dockerfile).

Reply:
All downloads here are now using curl and they are much better "structured"